packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/VideoRenderer.java

```diff
@@ -0,0 +1,326 @@
+              // to correct rotation
+              while (!Thread.interrupted()) {
+
+                synchronized (surfaceTextureAvailableFrameLock) {
+                  while (!surfaceTextureFrameAvailable) {
+                    surfaceTextureAvailableFrameLock.wait(500);
+                  }
+                  surfaceTextureFrameAvailable = false;
+                }
+
+                inputSurfaceTexture.updateTexImage();
+
+                float[] surfaceTextureMatrix = new float[16];
+                inputSurfaceTexture.getTransformMatrix(surfaceTextureMatrix);
+
+                draw(
+                    recordingWidth,
+                    recordingHeight,
+                    surfaceTextureMatrix,
+                    inputSurfaceTexture.getTimestamp());
+              }
+            } catch (InterruptedException e) {
+              Log.d(TAG, "thread interrupted while waiting for frames");
+            }
+          }
+        };
+    thread.start();
+  }
+
+  public int getTexId() {
+    return textureHandles[0];
+  }
+
+  public float[] moveMatrix() {
+    float[] m = new float[16];
+    Matrix.setIdentityM(m, 0);
+    Matrix.rotateM(m, 0, rotation, 0, 0, 1);
+    return m;
+  }
+
+  public void setRotation(int rotation) {
+    this.rotation = rotation;
+  }
+
+  private int loadShader(int type, String code) {
+
+    int shader = GLES20.glCreateShader(type);
+
+    GLES20.glShaderSource(shader, code);
+    GLES20.glCompileShader(shader);
+    return shader;
+  }
+
+  public void draw(int viewportWidth, int viewportHeight, float[] texMatrix, long timestamp) {
+
+    GLES20.glClear(GLES20.GL_COLOR_BUFFER_BIT | GLES20.GL_DEPTH_BUFFER_BIT);
+    GLES20.glClearColor(0f, 0f, 0f, 0f);
+
+    GLES20.glViewport(0, 0, viewportWidth, viewportHeight);
+
+    GLES20.glUseProgram(program);
+
+    // Pass transformations to shader
+    GLES20.glUniformMatrix4fv(texMatrixHandle, 1, false, texMatrix, 0);
+    GLES20.glUniformMatrix4fv(mvpHandle, 1, false, moveMatrix(), 0);
+
+    // Prepare buffers with vertices and indices & draw
+    GLES20.glBindBuffer(GLES20.GL_ARRAY_BUFFER, bufferHandles[0]);
+    GLES20.glBindBuffer(GLES20.GL_ELEMENT_ARRAY_BUFFER, bufferHandles[1]);
+
+    GLES20.glEnableVertexAttribArray(vertexHandle);
+    GLES20.glVertexAttribPointer(vertexHandle, 3, GLES20.GL_FLOAT, false, 4 * 5, 0);
+
+    GLES20.glEnableVertexAttribArray(uvsHandle);
+    GLES20.glVertexAttribPointer(uvsHandle, 2, GLES20.GL_FLOAT, false, 4 * 5, 3 * 4);
+
+    GLES20.glDrawElements(GLES20.GL_TRIANGLES, 6, GLES20.GL_UNSIGNED_INT, 0);
+
+    EGLExt.eglPresentationTimeANDROID(
+        display, surface, uptimeMillis() * 1000000); // Not the perfect solution but works well, see
+    // https://stackoverflow.com/questions/63467704/mediarecorder-surface-input-with-opengl-issue-if-audio-recording-is-enabled TODO: try on more devices and see if the timestamp can be incorporated
```

Comment:
Currently having no issue using the surface texture timestamp itself rather than system time. Which is how I would expect it to work. I have made those changes